### ISSUE TITLE
Get Endpoint for different URL

### ISSUE THREAD

**Issue by chucker** [dotnet/aspnetcore#18106]
### Is your feature request related to a problem? Please describe.
I am trying to get endpoint metadata for a given URL — basically a hypothetical `LinkGenerator.GetActionByUri()`;

I can inject `IActionDescriptorCollectionProvider` to get _all_ metadata, but I'd have to re-implement mapping the correct one (if any) based on templates. Before 2.2's Endpoints, the answer was apparently to create a `RouteContext` for your custom HttpContext, and take the first available `IRouteCollection` router:

```csharp
IRouteCollection router = RouteData.Routers.OfType<IRouteCollection>().First();
        
HttpContext context = _httpContextFactory.Create(HttpContext.Features);
context.Request.Path = "/Home/TestPostAction";
context.Request.Method = "POST";
```

However, `Routers` is now empty, [which is apparently by design](https://github.com/aspnet/AspNetCore/issues/5135#issuecomment-447733925).

It's not clear to me how to get an appropriate `IEndpointFeature` for a custom HttpContext. If I could _inject_ `IEndpointFeature`, I imagine something like the below would do:

```csharp
var featureCollection = new FeatureCollection();
featureCollection.Set<IEndpointFeature>(); // cannot inject this

var httpContext = _HttpContextFactory.Create(featureCollection);
httpContext.Request.Path = uri.PathAndQuery;

var feature = httpContext.Features.Get<IEndpointFeature>();
var endpoint = httpContext.Features.Get<IEndpointFeature>()?.Endpoint;
            
var metadata = endpoint.Metadata.OfType<Microsoft.AspNetCore.Mvc.Controllers.ControllerActionDescriptor>();
```

If I pass the _current_ HttpContext's `IEndpointFeature`, I get the metadata for the wrong endpoint (for obvious reasons).

### Additional context
This appears to be a feature regression compared to 2.1. See, for example, [this blog post](https://joonasw.net/view/find-out-if-url-matches-action).

**Comment by pranavkm** [comment]
You can get an `EndpointDataSource` instance from DI and read the list of endpoints in the application using - https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.routing.endpointdatasource.endpoints?view=aspnetcore-3.0#Microsoft_AspNetCore_Routing_EndpointDataSource_Endpoints. Does this work?

**Comment by chucker** [comment]
Thanks!

That works to get me _all_ endpoints, as you say, but won't match against a URL. However, those docs led me to the `LinkParser` class, which helps (and can be injected).

The following, while hardly efficient, appears to work:

```csharp
Endpoint? matchingEndpoint;

foreach (var endpoint in this._EndpointDataSource.Endpoints)
{
    if (!TryGetEndpointMetadata(endpoint, out var name, out var controllerActionDescriptor))
        continue;

    // we only use this to match the endpoint to the URL
    var dict = _LinkParser.ParsePathByEndpointName(name, forUri.AbsolutePath);
    if (dict == null)
        continue;

    // if we get here, it's the endpoint we're looking for
    matchingEndpoint = endpoint;
    break;
}
```

`TryGetEndpointMetadata` looks like:

```csharp
/// <summary>
/// Gets the routename and controller action descriptor for endpoints that have one.
/// </summary>
private bool TryGetEndpointMetadata(Endpoint endpoint, [NotNullWhen(true)] out string? name, [NotNullWhen(true)] out ControllerActionDescriptor? controllerActionDescriptor)
{
    controllerActionDescriptor = null;

    name = endpoint.Metadata.OfType<RouteNameMetadata>().FirstOrDefault()?.RouteName;

    if (name == null)
        return false;

    controllerActionDescriptor = endpoint.Metadata.OfType<ControllerActionDescriptor>().FirstOrDefault();

    return controllerActionDescriptor != null;
}
```

This allows me to:

1. pass any URL
2. if it matches a controller action, look at `controllerActionDescriptor.Parameters`

So it does solve my original problem, albeit inefficiently, It seems like there ought to be a nicer way of doing so than injecting a LinkParser and iterating all endpoints only to see if it returns a non-null value for one of them.

Basically, a `linkParser.MatchEndpointByPath()`?

Feel free to close this if this is the intended behavior.
